ecs-agent/acs/session/attach_instance_eni_responder.go

```diff
@@ -0,0 +1,154 @@
+
+	"github.com/aws/amazon-ecs-agent/ecs-agent/acs/model/ecsacs"
+	"github.com/aws/amazon-ecs-agent/ecs-agent/api/attachmentinfo"
+	apieni "github.com/aws/amazon-ecs-agent/ecs-agent/api/eni"
+	"github.com/aws/amazon-ecs-agent/ecs-agent/api/status"
+	"github.com/aws/amazon-ecs-agent/ecs-agent/logger"
+	"github.com/aws/amazon-ecs-agent/ecs-agent/logger/field"
+	"github.com/aws/amazon-ecs-agent/ecs-agent/wsclient"
+)
+
+const (
+	AttachInstanceENIMessageName = "AttachInstanceNetworkInterfacesMessage"
+)
+
+// attachInstanceENIResponder implements the wsclient.RequestResponder interface for responding
+// to ecsacs.AttachInstanceNetworkInterfacesMessage messages sent by ACS.
+type attachInstanceENIResponder struct {
+	eniHandler ENIHandler
+	respond    wsclient.RespondFunc
+}
+
+// NewAttachInstanceENIResponder returns an instance of the attachInstanceENIResponder struct.
+func NewAttachInstanceENIResponder(eniHandler ENIHandler,
+	responseSender wsclient.RespondFunc) wsclient.RequestResponder {
+	r := &attachInstanceENIResponder{
+		eniHandler: eniHandler,
+	}
+	r.respond = ResponseToACSSender(r.Name(), responseSender)
+	return r
+}
+
+func (*attachInstanceENIResponder) Name() string { return "attach instance ENI responder" }
+
+func (r *attachInstanceENIResponder) HandlerFunc() wsclient.RequestHandler {
+	return r.handleAttachMessage
+}
+
+func (r *attachInstanceENIResponder) handleAttachMessage(message *ecsacs.AttachInstanceNetworkInterfacesMessage) {
+	logger.Debug(fmt.Sprintf("Handling %s", AttachInstanceENIMessageName))
+	receivedAt := time.Now()
+
+	// Validate fields in the message.
+	if err := validateAttachInstanceNetworkInterfacesMessage(message); err != nil {
+		logger.Error(fmt.Sprintf("Error validating %s received from ECS", AttachInstanceENIMessageName), logger.Fields{
+			field.Error: err,
+		})
+		return
+	}
+
+	// Handle ENIs in the message.
+	messageID := aws.StringValue(message.MessageId)
+	for _, mENI := range message.ElasticNetworkInterfaces {
+		expiresAt := receivedAt.Add(time.Duration(aws.Int64Value(message.WaitTimeoutMs)) * time.Millisecond)
+		go func(eni *ecsacs.ElasticNetworkInterface) {
+			err := r.eniHandler.HandleENIAttachment(&apieni.ENIAttachment{
+				AttachmentInfo: attachmentinfo.AttachmentInfo{
+					TaskARN:              "",
+					AttachmentARN:        aws.StringValue(eni.AttachmentArn),
+					Status:               status.AttachmentNone,
+					ExpiresAt:            expiresAt,
+					AttachStatusSent:     false,
+					ClusterARN:           aws.StringValue(message.ClusterArn),
+					ContainerInstanceARN: aws.StringValue(message.ContainerInstanceArn),
+				},
+				AttachmentType: apieni.ENIAttachmentTypeInstanceENI,
+				MACAddress:     aws.StringValue(eni.MacAddress),
+			})
+			if err != nil {
+				logger.Error(fmt.Sprintf("Unable to handle %s", AttachInstanceENIMessageName), logger.Fields{
+					field.MessageID: messageID,
+					field.Error:     err,
+				})
+			}
+		}(mENI)
+	}
+
+	// Send ACK.
+	go func() {
```

Comment:
I see we do ack first then handle messages in previous implementation; I foresee no issue with current implementation, since message handling is done in go routine, but is there any reason we changed to this way?

Reply:
The main reason is to maintain consistency with other ACS message responders (i.e., right now, some responders we ACK first and others we ACK at the end with no clear reason as to why - we should look to standardize).

Reply:
Sounds good.